File: src/java/magma/app/locate/ParenthesesMatcher.java
package magma.app.locate;

import java.util.Optional;

public class ParenthesesMatcher implements Locator {
    @Override
    public String unwrap() {
        return ")";
    }

    @Override
    public int length() {
        return 1;
    }

    @Override
    public Optional<Integer> locate(String input) {
        var depth = 0;
        int i = 0;
        while (i < input.length()) {
            final var c = input.charAt(i);
            if (c == ')' && depth == 1) return Optional.of(i);
            if (c == '(') depth++;
            if (c == ')') depth--;
            i++;
        }
        return Optional.empty();
    }
}

File: src/c/magma/app/rule/locate/ParenthesesMatcher.c
#include "./ParenthesesMatcher.h"
struct ParenthesesMatcher implements Locator{
	String unwrap(){
		return ")";
	}
	int length(){
		return 1;
	}
	Stream<Integer> locate(String input){
		var depth=0;
		var queue=IntStream.range(0, input.length()).mapToObj(index->new Tuple<>(index, input.charAt(index))).collect(Collectors.toCollection(LinkedList::new));
		while(!queue.isEmpty()){
			var tuple=queue.pop();
			var i=tuple.left();
			var c=tuple.right();
			if(c=='\''){
				var tuple1=queue.pop();
				if(tuple1.right()=='\\'){
					queue.pop();
				}
				queue.pop();
			}
			if(c==')'&&depth==0)return Streams.of(i);
			if(c=='(')depth++;
			if(c==')')depth--;
		}
		return Streams.empty();
	}
}
